src/mscorlib/src/System/Diagnostics/Stacktrace.cs

```diff
@@ -532,23 +553,44 @@ internal String ToString(TraceFormat traceFormat)
             }
 
             bool fFirstFrame = true;
-            StringBuilder sb = new StringBuilder(255);
             for (int iFrameIndex = 0; iFrameIndex < m_iNumOfFrames; iFrameIndex++)
             {
                 StackFrame sf = GetFrame(iFrameIndex);
                 MethodBase mb = sf.GetMethod();
-                if (mb != null)
+                if (mb != null && (ShowInStackTrace(mb) || 
+                                   (iFrameIndex == m_iNumOfFrames - 1))) // Don't filter last frame
                 {
                     // We want a newline at the end of every line except for the last
                     if (fFirstFrame)
+                    {
+                        if (startNewLine)
+                        {
+                            sb.AppendLine();
+                        }
                         fFirstFrame = false;
+                    }
                     else
-                        sb.Append(Environment.NewLine);
+                    {
+                        sb.AppendLine();
+                    }
 
                     sb.AppendFormat(CultureInfo.InvariantCulture, "   {0} ", word_At);
 
                     Type t = mb.DeclaringType;
                     // if there is a type (non global method) print it
+
+                    if (t != null)
+                    {
+                        if (FormattingOptions.HasFlag(StackTraceFormattingOptions.ResolveAsyncMethods) && typeof(IAsyncStateMachine).IsAssignableFrom(t))
+                        {
+                            t = ResolveAsyncMethod(t, ref mb);
+                        }
+                        else if (FormattingOptions.HasFlag(StackTraceFormattingOptions.ResolveIteratorMethods) && typeof(IEnumerable).IsAssignableFrom(t))
```

Comment:
Should check for `IEnumerator` instead (works if iterator returns `IEnumerable` or `IEnumerator`)

Reply:
Good tip